### PR TITLE
Move caching logic to response

### PR DESCRIPTION
**Pull request by rullzer**
This avoids having to do it at all the places we want cached responses.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

**Comment by rullzer** [comment]
I didn't tackle the immutable directive yet. Maybe it makes sense to just drop it. Support is very limited anyways: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Browser_compatibility

Also if we set the ttl to a year it should be good enough already.

File: lib/public/AppFramework/Http/Response.php [hunk 1]
```diff
@@ -95,12 +96,23 @@ class Response {
...
-	public function cacheFor($cacheSeconds) {
-
+	public function cacheFor(int $cacheSeconds) {
 		if($cacheSeconds > 0) {
 			$this->addHeader('Cache-Control', 'max-age=' . $cacheSeconds . ', must-revalidate');
+
+			// Old scool prama caching
+			$this->addHeader('Pragma', 'public');
+
+			// Cache previews for 24H
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/Response.php]: what previews? :wink:

File: lib/public/AppFramework/Http/Response.php [hunk 2]
```diff
@@ -95,12 +96,23 @@ class Response {
...
+			$expires = new \DateTime();
+			/** @var ITimeFactory $time */
+			$time = \OC::$server->query(ITimeFactory::class);
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/Response.php]: if you query the time factory this way you can also just use `time()` :see_no_evil:

Comment (ChristophWurst) [reply lib/public/AppFramework/Http/Response.php]: I wonder if that will :boom: for some tests as expected and actual response might have different time stamps set for their expiration dates (e.g. if you construct a response in your test code to compare to the actual response).

Comment (ChristophWurst) [reply lib/public/AppFramework/Http/Response.php]: FYI my approach on this for the mail app looks like this: https://github.com/nextcloud/mail/blob/0b8f0490146438ab3a4c39ec02e993a2565e7528/lib/Http/CacheHeaders.php. It's a trait that adds a method which takes a `ITimeFactory` instance. This way I can re-use it for any custom response and also mock the time if necessary.

Comment (rullzer) [reply lib/public/AppFramework/Http/Response.php]: Actually doing it like this does allow us to unit test it. As you can overwrite services.

Comment (ChristophWurst) [reply lib/public/AppFramework/Http/Response.php]: ah right, forgot about that :+1:

**Review by ChristophWurst — APPROVED** [review]
Nice!

**Review by MorrisJobke — APPROVED** [review]

**Pull request merged**